File: frontend/ui/src/app/types/subscription.ts
export const UNLIMITED_QTY = -1;
export type SubscriptionType = 'community' | 'starter' | 'pro' | 'enterprise' | 'trial';

export type SubscriptionPeriod = 'monthly' | 'yearly';

export interface SubscriptionLimits {
  maxCustomerOrganizations: number;
  maxUsersPerCustomerOrganization: number;
  maxDeploymentsPerCustomerOrganization: number;
}

export interface SubscriptionInfo {
  subscriptionType: SubscriptionType;
  subscriptionPeriod: SubscriptionPeriod;
  subscriptionEndsAt: string;
  subscriptionCustomerOrganizationQuantity: number;
  subscriptionUserAccountQuantity: number;
  currentUserAccountCount: number;
  currentCustomerOrganizationCount: number;
  currentMaxUsersPerCustomer: number;
  currentMaxDeploymentTargetsPerCustomer: number;
  hasApplicationLicenses: boolean;
  hasArtifactLicenses: boolean;
  hasNonAdminRoles: boolean;
  hasAlertConfigurations: boolean;
  limits: {[key in SubscriptionType]: SubscriptionLimits};
}

export interface CheckoutRequest {
  subscriptionType: SubscriptionType;
  subscriptionPeriod: SubscriptionPeriod;
  subscriptionUserAccountQuantity: number;
  subscriptionCustomerOrganizationQuantity: number;
}

File: api/subscription.go
package api

import (
	"time"

	"github.com/distr-sh/distr/internal/types"
)

type CheckoutResponse struct {
	SessionID string `json:"sessionId"`
	URL       string `json:"url"`
}

type SubscriptionLimits struct {
	MaxCustomerOrganizations        int64 `json:"maxCustomerOrganizations"`
	MaxUsersPerCustomerOrganization int64 `json:"maxUsersPerCustomerOrganization"`
	MaxDeploymentsPerCustomerOrg    int64 `json:"maxDeploymentsPerCustomerOrganization"`
}

type SubscriptionInfo struct {
	SubscriptionType                       types.SubscriptionType   `json:"subscriptionType"`
	SubscriptionPeriod                     types.SubscriptionPeriod `json:"subscriptionPeriod"`
	SubscriptionEndsAt                     time.Time                `json:"subscriptionEndsAt"`
	SubscriptionCustomerOrganizationQty    int64                    `json:"subscriptionCustomerOrganizationQuantity"`
	SubscriptionUserAccountQty             int64                    `json:"subscriptionUserAccountQuantity"`
	CurrentUserAccountCount                int64                    `json:"currentUserAccountCount"`
	CurrentCustomerOrganizationCount       int64                    `json:"currentCustomerOrganizationCount"`
	CurrentMaxUsersPerCustomer             int64                    `json:"currentMaxUsersPerCustomer"`
	CurrentMaxDeploymentTargetsPerCustomer int64                    `json:"currentMaxDeploymentTargetsPerCustomer"`
	HasApplicationLicenses                 bool                     `json:"hasApplicationLicenses"`
	HasArtifactLicenses                    bool                     `json:"hasArtifactLicenses"`
	HasNonAdminRoles                       bool                     `json:"hasNonAdminRoles"`
	HasAlertConfigurations                 bool                     `json:"hasAlertConfigurations"`

	Limits map[types.SubscriptionType]SubscriptionLimits `json:"limits"`
}
